### ISSUE TITLE
Remove normalization in `UPGrad` and `DualProj`

### ISSUE THREAD

**Issue by PierreQuinton** [TorchJD/torchjd#242]
There is no need for normalization and it slows down the computation. Note that this requires a change to tests to not test very large matrices.

**Comment by ValerianRey** [comment]
We need to be sure that this doesn't make UPGrad or DualProj worse. It's also a breaking change, so we will have to document the change nicely in the changelog.

Apart from that, I really agree that it would be better: it's one less hyper-parameter, and one less linear algebra operation to make (computing the svd), so it will make our algorithm faster, easier to re-implement, and easier to maintain.

**Comment by PierreQuinton** [comment]
It should not make UPGrad or DualProj in moderate regimes I would say, because in some regimes it will have nans which is in itself unavoidable anyways (except in some extremely specific cases such as the one of our tests).

**Comment by ValerianRey** [comment]
I'm starting a PR for this.
A few things to note already:
- CAGrad currently also depends on the normalized Gramian, so this will not affect just UPGrad and DualProj.
- I think that without normalization, the regularization matrix should be multiplied by the maximum singular value of the Gramian, in order to preserve (more or less) the same behavior as before. If we don't do that, the obtained Gramian is not PD for fairly large matrices (even 10^5 scale), so this seems mandatory. Maybe we can multiply by something else than the largest singular value, but we need to multiply by something that depends on the norm of the matrix.

**Comment by PierreQuinton** [comment]
I have experimented with this on my side, `UPGrad` and `DualProj` then typically lead to quadprog saying that the gramian is not PSD, it could be due to:
- lack of symmetry (can be fixed by taking $1/2(G+G^\top)$
- bad condition number
- nans

**Comment by ValerianRey** [comment]
> Can be fixed by taking $1/2 (G + G^T)$

Do you suggest to replace the regularization by this "symmetrization", or do you think we should have both?

> and alternative to what you propose is to provide a different reg_eps in the tests based on the matrix, we could also document that in upgrad.

I don't like that so much: users shouldn't have to adapt the reg_eps based on the Jacobian norm. Imagine if the Jacobian reaches zero towards the end of the training. I'm pretty sure that the reg_eps would need to also be reduced towards zero in this case. This is not practical at all. It's also not easy to implement with the current structure of property testers.

> I am also working on an implementation of those not using the Gramian, but rather the QR decomposition. quadprog has this parameter factorized that can be used to provide another matrix that we could in principle compute without even computing the Gramian. I will make a note on that to your intention.

That's an interesting direction! Please keep me updated.

**Comment by ValerianRey** [comment]
Other than that, would it be possible to compute the maximum singular value of G without too many computations? I think that would be much easier, and it would require less heavy testing to check that we're not breaking UPGrad and DualProj.

**Comment by ValerianRey** [comment]
> Other than that, would it be possible to compute the maximum singular value of G without too many computations? I think that would be much easier, and it would require less heavy testing to check that we're not breaking UPGrad and DualProj.

I tried this and quadprog still fails. Note that it returns None, so it's not due to the matrix being non-PSD, but just from the matrix values being too large I guess.

So I think we can't get rid of normalization before we change the solver.

**Comment by ValerianRey** [comment]
Closing this issue since we can't get rid of normalization unless we stop using quadprog. We should reassess this if/when we start using our own torch-based qp solver.
